### PR TITLE
feat: add offline banner

### LINKED ISSUE #194: Link the whole new offline logic to UI

### PR DESCRIPTION
**Pull request by rami-mallah**
> DISCLAIMER: This PR shows that there was 1100 additions and 600 deletions, but there is _only 120 lines_ of new code and implementation. Only 10 lines of code are relevant in the `AppNavHost` because I wrapped the whole code inside a `Column`, so you don't need to review everything inside the `Column` (700 additions and the 600 deletions are not relevant). The rest of the lines in other files are documentations and tests.

## Summary

This PR implements an offline banner feature that displays at the top of all screens when the device is offline. The banner shows "You're offline" along with the last successful data synchronization time (e.g., "Last updated 2 hours ago"). This provides users with clear feedback about their connection status and how fresh their cached data is.
This closes #194.

## Features

- Offline Detection: Banner automatically appears when network is unavailable and hides when connection is restored
- Last Sync Time: Displays when data was last successfully synchronized (e.g., "Last updated 2 hours ago")
- No Sync Indicator: Shows "No sync data available" if the app has never synced data
- Automatic Updates: Banner state updates every 2 seconds to reflect current network status
- Global Display: Banner appears at the top of all screens via integration in `AppNavHost`

### 1. LastSyncTracker Utility (`LastSyncTracker.kt`)
- Purpose: Tracks the timestamp of the last successful data synchronization
- Storage: Uses `SharedPreferences` to persist sync timestamps across app restarts
  
### 2. OfflineBanner Composable (`OfflineBanner.kt`)
- Polls network state every 2 seconds using `LaunchedEffect`
- Only displays when `NetworkUtils.isNetworkAvailable()` returns `false`

### 3. Repository Integration
Both hybrid repositories now record sync timestamps after successful data synchronization:
- `ReviewsRepositoryHybrid.kt` and `RentalListingRepositoryHybrid.kt` record sync after `syncToLocal()` completes
This ensures the banner always shows the most recent sync time for any data type.

### 4. UI Integration (`AppNavHost.kt`)
- Wrapped `NavHost` in a `Column` to place the banner above all screens
- Added `OfflineBanner()` at the top of the column
- Used `Modifier.weight(1f)` on `NavHost` to ensure it takes remaining space

## Testing

Unit tests were added for `LastSyncTracker` to verify timestamp recording, retrieval, and edge case handling. 
UI tests were added for `OfflineBanner` to verify banner visibility, message display, and dynamic updates based on network state.

MockK was used in the UI tests to mock `NetworkUtils` because we need to test the banner's behavior when offline/online, but we cannot reliably control the actual device's network state in tests. By mocking the network state, we ensure tests are deterministic and can simulate different network scenarios (online/offline transitions) programmatically without requiring root access or complex system-level changes.

Screenshot_20251203_185839

## Notes
- All documentation was generated with the help of AI. This PR comment was refined by AI, to improve the explanations and to remove the repetition of "I did" and writing in a non personal way.
- Although SonarCloud shows a line coverage of 79.3%, the line that are not covered are actually lines from `AppNavHost` that aren't added by this PR, but as explained before, they counted as "new code". The only lines of logic added to `AppNavHost` are labeled as tested.

### CODE REVIEW

File: app/src/main/java/com/android/mySwissDorm/utils/LastSyncTracker.kt [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    } else {
+      null
+    }
+  }
+
+  /**
+   * Clears the stored last sync timestamp.
+   *
+   * This can be useful for testing or when resetting the app state.
+   *
```
Comment (ghassanbaroudi) [inline app/src/main/java/com/android/mySwissDorm/utils/LastSyncTracker.kt]: Looks good and clean, welld one.

**Review by ghassanbaroudi — APPROVED** [review]
The implementation is great and looks to be working well when checking out on your branch. It is smooth, and I don't see any issues with it. I like the fact that it displays the last sync time too. 
The tests are long and cover all the cases I could think of, which is good. Thank for this new feature!
